src/core/ledger/currency.js

```diff
@@ -0,0 +1,79 @@
+// @flow
+
+import {
+  ethAddressParser,
+  type EthAddress,
+} from "../../plugins/ethereum/ethAddress";
+import * as C from "../../util/combo";
+
+/**
+ * EvmChainId is represented in number form for all EVM-based chains,
+ * including mainnet (1), and xDai (100)
+ * The reason for this is that ethereum's client configuration utilizes
+ * a number to represent chainId, and this way we can just transpose that
+ * chainId here as a component of the currency Id, since the web3 client will
+ * return a stringified integer when the chainId is requested.
+ */
+export type EvmChainId = string;
+
+export function parseEvmChainId(id: string): EvmChainId {
+  const result = parseInt(id, 10).toString();
+  if (Number.isNaN(result) || result !== id) {
+    throw new Error(`Invalid EVM chainId value: ${id}`);
+  }
+  return id;
+}
+
+export const evmChainParser: C.Parser<EvmChainId> = C.fmap(
+  C.string,
+  parseEvmChainId
+);
+
+declare type EvmId = {
+  chainId: EvmChainId,
+  /**
+   * CurrencyAddress is a subset of all available EthAddresses.
+   *
+   * A currency address is the address of the token contract for an ERC20 token,
+   * or the 20 byte-length equivalent of 0x0, which is the conventional address
+   * used to represent ETH on the ethereum mainnet, or the native currency on an
+   * EVM-based sidechain see here for more details on these semantics:
+   * https://ethereum.org/en/developers/docs/intro-to-ethereum/#eth
+   */
+  currencyAddress: EthAddress,
+};
+
+export const evmIdParser: C.Parser<EvmId> = C.object({
+  chainId: evmChainParser,
+  currencyAddress: ethAddressParser,
+});
+
+/**
+ * Example protocol symbols: "BTC" for bitcoin and "FIL" for Filecoin
+ */
+declare type ProtocolSymbol = string;
+
+/**
+ * Chains like Bitcoin and Filecoin do not have "production" sidechains so
+ * we represent them as a string, as specified in the ProtocolSymbol type
+ */
+declare type ProtocolId = {
+  chainId: ProtocolSymbol,
+  // currencyAddress is unused here
+  currencyAddress?: any,
+};
+
+export const protocolIdParser: C.Parser<ProtocolId> = C.object(
+  {
+    chainId: C.exactly(["BTC", "FIL"]),
+  },
+  {
+    currencyAddress: C.pure<void>(),
+  }
+);
+
+export type ChainId = ProtocolSymbol | EvmChainId;
+export type CurrencyId = ProtocolId | EvmId;
```

Comment:
ProtocolSymbol being assigned to chainId is a bit confusing. Feel like it might be cleaner to have a single type for all tokens / currencies, something like:

```ts
declare type Currency = {
  symbol: string,
  chainId?: EvmChainId,
  tokenAddress?: EvmAddress
}
```

Where `chainId` and `tokenAddress` would be undefined for non-EVM currencies.

Reply:
@hammadj I considered this monolithic type but I thought is was more confusing for static typing. By splitting between these two subtypes, it's clearer in flow what type we're expecting and what sort of details it should possess.

Reply:
Instead of doing the Union at the ID level, could just do something like this:

```ts

type EvmToken = {
  symbol: string,
  chainId: EvmChainId,
  tokenAddress: EvmAddress
}

type ProtocolToken = {
  symbol: string,
}

type Currency = EvmToken | ProtocolToken
```

This would also allow us to add additional fields scoped to different protocol tokens (e.g. if Polkadot has something like ERC20's on their network then we could have a tokenAddress field on ProtocolToken as well but enforce it to follow the semantics of its protocol instead of allowing a EvmChainId to be paired with a ProtocolId type)